EasyThreeD ET4000+ Mainboard

schmttc opened the pull request:
### Description

Port of a manufacturer’s version of Marlin 1.1.1 to Marlin 2.0 for the EasyThreeD K7 printer, mainboard ET4000+.
The original was a ‘dirty’ implementation where the standard Marlin structure was not used, and hardware specific features were spliced into places they don’t belong.
- The button input model includes a slide switch for feed/retract filament (ie. behaviour based on state change); and time-length button holds to perform different tasks using a single button (raise print head; and start print). The standard Marlin button input model only supports a single press for button input.
- There is also a status LED that blinks/solid lights for various printer status
- The physical pins for these buttons and LED are the ones typically used to connect an LCD module
- No display

I have abstracted the code where possible, as EasyThreeD make another printer the X1 which has a different configuration of screenless, button only input. So someone else can edit these files to port in this other model at some point.

### Requirements

- Build with PIO board 'mks_robin_lite_maple'
- EasyThreeD K7 Printer (more recent models that shipped with the ET4000+ mainboard)
- Should also work with the EasyThreeD Nano

### Benefits
Implements standard Marlin firmware for the EasyThreeD ET4000+ mainboard 

### Configurations
[marlin.config.zip](https://github.com/MarlinFirmware/Marlin/files/7481128/marlin.config.zip)

### Related Issues

mkrelle committed:
EasyThreeD 4000+ Mainboard

thinkyhead committed:
apply coding standards

thinkyhead committed:
setPins => init

thinkyhead committed:
define exp1 header (best guess)

thinkyhead commented:
This just needs some of the `EasythreedUI` to be brought up to date with the latest techniques for the actions it initiates. It's not clear how the UI is supposed to work for starting an SD file, and so on. Can you post a video or link to a video that demonstrates how all the EasythreedUI operations actually work?

schmttc commented:
Thanks, I really appreciate your effort to clean up my integration attempts.
I'll put a video together in the next few days using the stock firmware to demonstrate.

Regarding the EXP1 port for LCD, it is 14 pins, unlabelled. This is the same as the MKS Robin Lite and MKS Robin Lite 2 boards, eg https://www.aliexpress.com/item/32894328196.html
I'm not able to find a documented pinout, but can trace the known pins for the EasyThreeD K7 in the next week or so if that helps?

schmttc committed:
Comments, Pause/Resume function

schmttc commented:
Here's a video of the user interface on the stock firmware https://drive.google.com/file/d/15h1T7LJbTH4cI1J6UYHpgWyh7p-XdCE0/view
I've updated the code to include the pause/resume function that is also present in the original https://github.com/schmttc/EasyThreeD-K7-STM32/blob/Original-Firmware-Marlin-1.1.1/mksRobinLite_nano/Src/nano.cpp
But on testing, the 'Pause' doesn't currently work.

thinkyhead committed:
whitespace

thinkyhead committed:
Various adjustments and commentary

thinkyhead commented on Marlin/src/feature/easythreed_ui.cpp:
```diff
@@ -0,0 +1,242 @@
+              card.ls();                                            // List all files to serial output
+              const uint16_t filecnt = card.countFilesInWorkDir();  // Count printable files in cwd
+              if (filecnt == 0) return;                             // None are printable?
+              card.selectFileByIndex(filecnt);                      // Select the last file according to current sort options
```
Since file sorting may be alphabetical, reversed, etc., the actual file selected by this method is unpredictable.

schmttc replied:
The intention is to select the most recent file by date/time from the SD card. With the current code this works; but is there a preferred method to achieve this outcome?

thinkyhead commented on Marlin/src/feature/easythreed_ui.cpp:
```diff
@@ -0,0 +1,242 @@
+              print_key_flag = PF_PAUSE;                            // The "Print" button now pauses the print
+            }
+            break;
+          case PF_PAUSE:                                            // Pause printing (not currently firing)
```
If this is not firing, look for where `print_key_flag` is being changed incorrectly.

thinkyhead commented on Marlin/src/feature/easythreed_ui.cpp:
```diff
@@ -0,0 +1,242 @@
+            break;
+          case PF_PAUSE:                                            // Pause printing (not currently firing)
+            blink_interval_ms = LED_ON;                             // Set indicator to steady ON
+            card.pauseSDPrint();                                    // Unceremoniously pause the SD print in progress
```
It is better to send `M25` because this only pauses the SD card processing. It doesn't park, etc.

thinkyhead commented on Marlin/src/feature/easythreed_ui.cpp:
```diff
@@ -0,0 +1,242 @@
+            break;
+          case PF_RESUME:                                           // Resume printing (not currently firing)
+            blink_interval_ms = LED_BLINK_2;                        // Blink the indicator LED at 1 second intervals
+            card.startOrResumeFilePrinting();                       // Unceremoniously resume the SD print in progress
```
It is better to send `M24` because this only resumes SD card processing. It doesn't un-park, etc.

thinkyhead commented on Marlin/src/feature/easythreed_ui.cpp:
```diff
@@ -0,0 +1,242 @@
+          blink_interval_ms = LED_OFF;                              // Turn off LED
+        }
+        planner.synchronize();                                      // Wait for commands already in the planner to finish
+        TERN_(HAS_STEPPER_RESET, disableStepperDrivers());          // Disable all steppers - now!
```
Note that the commands sent by `inject` above will be executed later, after exiting this method.

thinkyhead replied:
Equivalent to… `"G91\nG0 Z10 F600\nG90\nM400\nM18"`

thinkyhead commented on Marlin/src/feature/easythreed_ui.cpp:
```diff
@@ -0,0 +1,242 @@
+          quickstop_stepper();                                      // Hard-stop all steppers
+          thermalManager.disable_all_heaters();                     // Turn off all heaters
+          print_flag = false;
+          blink_interval_ms = LED_OFF;                              // Turn off LED
```
Nothing in this block will actually abort the print in progress. The correct method to abort a print is to send `M524` or call `card.abortFilePrintSoon()` (when `IS_SD_PRINTING()`).

thinkyhead commented on Marlin/src/feature/easythreed_ui.cpp:
```diff
@@ -0,0 +1,242 @@
+        flag = false;                                               // Restore flag to false
+        filament_status = FS_IDLE;                                  // Go back to idle state
+        quickstop_stepper();                                        // Hard-stop all the steppers ... now!
+        thermalManager.disable_all_heaters();                       // And disable all the heaters
```
The feed / retract methodology employed here is … interesting. Also, it is somewhat likely to corrupt the `current_position`. There is a possibility that the machine may be in motion, or in the middle of `G29`, or otherwise not idle when this code is processed.

thinkyhead commented:
I cleaned up the code, added comments, and did some review on the current code. Some corrections are still needed. The trick of aborting feed/retract moves in progress on button release may interact badly with active moves, so perhaps E can be moved more directly, with shorter segments, checking `planner.movesplanned()` before adding additional small moves.

schmttc committed:
Updates and Comments

schmttc commented:
- It looks like a couple of lines in the print case aren't executing, but I don't understand why https://github.com/schmttc/Marlin/commit/f8e20bf932e772588567025b3de5be3e3168ebe5#r60337555 
- Added some comments; in particular the Load/Unload "buttons" are a 3 position switch with a common center ground, which explains the 'both released' query
- Updated card.abortFilePrintSoon(); as the cancel function
- Fixed LED state exiting after filament advance

schmttc committed:
Buttons all working

schmttc committed:
Buttons all working + original config files

schmttc commented:
Completed:
-All buttons trigger when expected
-LED functions as expected
-Use M24/M25 for pause/resume

To Do:
-Don't allow pause/resume while pre-heating
-Use M400 and M18 instead of planner.synchronize() and TERN_(HAS_STEPPER_RESET, disableStepperDrivers());    
-Cancel Print: Use M524 or card.abortFilePrintSoon() (when IS_SD_PRINTING())
-Fix feed/retract E stop concern
-Fix potentially unpredictable file sort

thinkyhead committed:
fix formatting

schmttc commented:
Thank you for merging. I will create a new PR to address the outstanding items when ready.
